examples/module/controller_parameters_tuner/multicoptor_controller_tuner.py

```diff
@@ -0,0 +1,219 @@
+
+      system_state = system_new_state
+
+      loss += torch.norm(
+        ref_position - position
+      )
+    return loss / len(ref_states)
+
+
+def func_to_get_state_error(state, ref_state):
+    ref_position, ref_velocity, ref_acceleration, \
+      ref_pose, ref_angular_vel, ref_angular_acc = ref_state
+
+    return state - torch.concat(
+       [
+          ref_position,
+          torch.squeeze(torch.t(rotation_matrix_2_quaternion(ref_pose))),
+          ref_velocity,
+          ref_angular_vel
+       ]
+    )
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser(description='Cartpole Example')
+    parser.add_argument("--device", type=str, default='cpu', help="cuda or cpu")
+    parser.add_argument("--save", type=str, default='./examples/module/dynamics/save/',
+                        help="location of png files to save")
+    parser.add_argument('--show', dest='show', action='store_true',
+                        help="show plot, default: False")
+    parser.set_defaults(show=False)
+    args = parser.parse_args(); print(args)
+    os.makedirs(os.path.join(args.save), exist_ok=True)
+
+    g = 9.81
+
+    # program parameters
+    time_interval = 0.05
+    learning_rate = 0.01
+    initial_state = torch.tensor([
+        torch.tensor(0.), # position.x
+        torch.tensor(0.), # position.y
+        torch.tensor(0.), # position.z
+        torch.tensor(1.), # quaternion.w
+        torch.tensor(0.), # quaternion.x
+        torch.tensor(0.), # quaternion.y
+        torch.tensor(0.), # quaternion.z
+        torch.tensor(0.), # vel.x
+        torch.tensor(0.), # vel.y
+        torch.tensor(0.), # vel.z
+        torch.tensor(0.), # anguler_vel.x
+        torch.tensor(0.), # anguler_vel.y
+        torch.tensor(0.)] # anguler_vel.z
+      , device=args.device).double()
+    initial_controller_parameters = torch.tensor([
+        torch.tensor(1.),
+        torch.tensor(1.),
+        torch.tensor(1.),
+        torch.tensor(1.)]
+      , device=args.device).double()
+
+    points = torch.tensor([[[0., 0., 0.],
+                            [1., 1., -2],
+                            [3., 3., -4]]], device=args.device)
+
+    waypoints = pp.CSplineR3(points, time_interval)[0]
+
+    ref_states = get_ref_states(initial_state, waypoints, time_interval)
+
+    e3 = torch.stack([
+        torch.tensor([0.]),
+        torch.tensor([0.]),
+        torch.tensor([1.])]
+      ).to(device=args.device).double()
+    multicopter = MultiCopter(time_interval,
+                               0.6,
+                               torch.tensor(g),
+                               torch.tensor([
+                                  [0.0829, 0., 0.],
+                                  [0, 0.0845, 0],
+                                  [0, 0, 0.1377]
+                                ], device=args.device),
+                                e3)
+
+    # start to tune the controller parameters
+    max_tuning_iterations = 30
```

Comment:
The paper uses a more detailed termination condition, such as loss, instead of hard coded iterations